Add QElidingLabel

tlambert03 opened the pull request:
closes #14 

Adds a `QLabel` variant that elides text to fill the available space.  For multi-line Qlabel (`setWordWrap(True)`), it will truncate the number of lines shown, and elide the last line, to fit the available space:

```py
In [1]: from superqt import QElidingLabel

In [2]: text = '''Lorem ipsum dolor sit amet, consectetur adipiscing elit, 
   ...: sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 
   ...: Ut enim ad minim ven iam, quis nostrud exercitation ullamco laboris
   ...: nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
   ...: reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.'''

In [3]: wdg = QElidingLabel(text)

In [4]: wdg.resize(260, 80)

In [5]: wdg.show()
```
<img width="372" alt="Untitled" src="https://user-images.githubusercontent.com/1609449/128371190-ff6a4784-b68d-4b75-abd3-0626d9cba1bf.png">

```py
In [6]: wdg.setWordWrap(True)
```

<img width="372" alt="Untitled 2" src="https://user-images.githubusercontent.com/1609449/128371184-e66e0455-c12f-4eff-b4d7-1e9a3a4cd506.png">

tlambert03 commented:
@goanpeca, @Czaki, @sofroniewn ... if any of you have a moment to take a quick peek here, i'd appreciate it.  would prefer not to merge anything in this repo without at least one other set of eyes. :)

Czaki commented:
Maybe set a tooltip with full text?

Czaki commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+            return fm.elidedText(self._text, self._elide_mode, width)
+
+        # get number of lines we can fit without eliding
+        nlines = self.height() // fm.height() - 1
```
There is no pass of new height (for resize event) so there is a possibility for a wrong number of lines.

tlambert03 replied:
according to the [resizeEvent docs](https://doc.qt.io/qt-5/qwidget.html#resizeEvent)

> When resizeEvent() is called, the widget already has its new geometry.

so `self.height()` should already be correct.  really, I didn't need to pass width from the resize event either. perhaps I should remove that?

sofroniewn reviewed (approved):
Didn't try running this, but gave it a quick look over and code looks good

Czaki commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+        self._elide_mode = Qt.TextElideMode(mode)  # type: ignore
+        super().setText(self._elidedText())
+
+    def fullText(self) -> str:
+        """Return the full un-elided text."""
+        return self._text
```
why this instead of plain `text`?

goanpeca replied:
Same question 🤔

tlambert03 replied:
great question both :) I thought I wasn't going to be able to do this since I was using the internal model to hold the elided text, but it works out fine.  thanks!

goanpeca commented on examples/eliding_label.py:
```diff
@@ -0,0 +1,12 @@
+
+app = QApplication([])
+
+labl = QElidingLabel(
```
Very minor nitpick, but using `label` feels better than `labl` or `lbl` or `l` :)

goanpeca commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+    full text.
+    """
+
+    ELIDE_STRING = "…"
```
Should this be private?

Do we want to be able to configure this? is this locale related?

tlambert03 replied:
removed

goanpeca commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+        return self._text
+
+    @staticmethod
+    def wrapText(text, width, font=QFont()) -> List[str]:
```
This feels weird `font=QFont()` 

Why not use a `None` ?

tlambert03 replied:
can do,  (have to convert to an empty font for the call to QTextLayout, but can do `QTextLayout(text, font or QFont())`)

goanpeca commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+
+    # Reimplemented QT methods
+
+    def setText(self, txt):
```
what about text?

goanpeca replied:
txt -> string, text, value

tlambert03 replied:
added

goanpeca commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+        self._text = txt
+        super().setText(self._elidedText())
+
+    def resizeEvent(self, rEvent: QResizeEvent):
```
rEvent -> event ?

goanpeca commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+        super().setText(self._elidedText(rEvent.size().width()))
+        rEvent.accept()
+
+    def setWordWrap(self, wrap: bool) -> None:
```
Docsrting?

goanpeca commented on superqt/_eliding_label.py:
```diff
@@ -0,0 +1,105 @@
+        super().setWordWrap(wrap)
+        super().setText(self._elidedText())
+
+    def sizeHint(self):
```
Docstring?

tlambert03 replied:
in looking into adding these, it turns out that even the `super()` implementation didn't have a docstring, and I don't want to start creating custom docstrings for builtin methods.  currently it gives the same result as the native method which works for me

```py
In [10]: QLabel.sizeHint?
Signature: QLabel.sizeHint(self) -> PySide2.QtCore.QSize
Docstring: sizeHint(self) -> PySide2.QtCore.QSize
Type:      method_descriptor

In [11]: QElidingLabel.sizeHint?
Signature: QElidingLabel.sizeHint(self) -> PySide2.QtCore.QSize
Docstring: sizeHint(self) -> PySide2.QtCore.QSize
```

goanpeca replied:
👍🏼

goanpeca reviewed (changes requested):
Hi @tlambert03 thanks for working on this.

Made some requests and comments.

Overall looks good!

tlambert03 commented:
thank you all!  @goanpeca @Czaki, I agree with the `fullText` to `text()` comments.  The reason I wasn't using text is, as you probably noticed, I'm actually using the public QLabel `text` as the elided text.  (so `text()` would return the elided version).  But I'm convinced by your comments that that's not the right approach, and will refactor to put that logic into the paintEvent rather than the model itself (which is where it belongs anyway)

tlambert03 commented:
I think I addressed everything, and added full test coverage.  Will merge later today unless anyone has more requests. 

will then cut a new release with this and @Czaki's new `QEnumComboBox`.

goanpeca reviewed (approved):
🚀
